// @openmrs/esm-react-utils/dist/ExtensionSlot.js
/** @module @category Extension */ import React, { useRef, useMemo } from "react";
import { ComponentContext } from "./ComponentContext.js";
import { Extension } from "./Extension.js";
import { useExtensionSlot } from "./useExtensionSlot.js";
function defaultSelect(extensions) {
    return extensions;
}
/**
 * An [extension slot](https://o3-docs.openmrs.org/docs/extension-system).
 * A place with a name. Extensions that get connected to that name
 * will be rendered into this.
 *
 * @example
 * Passing a react node as children
 *
 * ```tsx
 * <ExtensionSlot name="Foo">
 *   <div style={{ width: 10rem }}>
 *     <Extension />
 *   </div>
 * </ExtensionSlot>
 * ```
 *
 * @example
 * Passing a function as children
 *
 * ```tsx
 * <ExtensionSlot name="Bar">
 *   {(extension) => (
 *     <h1>{extension.name}</h1>
 *     <div style={{ color: extension.meta.color }}>
 *       <Extension />
 *     </div>
 *   )}
 * </ExtensionSlot>
 * ```
 *
 */ export function ExtensionSlot({ name: extensionSlotName, extensionSlotName: legacyExtensionSlotName, select = defaultSelect, children, state, style, ...divProps }) {
    if (children && state) {
        throw new Error('Both children and state have been provided. If children are provided, the state must be passed as a prop to the `Extension` component.');
    }
    const name = extensionSlotName ?? legacyExtensionSlotName;
    const slotRef = useRef(null);
    const { extensions, extensionSlotModuleName } = useExtensionSlot(name, state);
    const extensionsToRender = useMemo(()=>select(extensions), [
        select,
        extensions
    ]);
    const extensionsFromChildrenFunction = useMemo(()=>{
        if (typeof children == 'function' && !/*#__PURE__*/ React.isValidElement(children)) {
            return extensionsToRender.map((extension)=>children(extension, state));
        }
    }, [
        children,
        extensionsToRender
    ]);
    return /*#__PURE__*/ React.createElement("div", {
        ref: slotRef,
        "data-extension-slot-name": name,
        "data-extension-slot-module-name": extensionSlotModuleName,
        style: {
            ...style,
            position: 'relative'
        },
        ...divProps
    }, name && extensionsToRender?.map((extension, i)=>/*#__PURE__*/ React.createElement(ComponentContext.Provider, {
            key: extension.id,
            value: {
                moduleName: extensionSlotModuleName,
                featureName: '',
                extension: {
                    extensionId: extension.id,
                    extensionSlotName: name,
                    extensionSlotModuleName
                }
            }
        }, extensionsFromChildrenFunction?.[i] ?? (typeof children !== 'function' ? children : null) ?? /*#__PURE__*/ React.createElement(Extension, {
            state: state
        }))));
}
