// @openmrs/esm-react-utils/dist/openmrsComponentDecorator.js
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
import React, { Suspense } from "react";
import { I18nextProvider } from "react-i18next";
import { SWRConfig } from "swr";
import { openmrsFetch, OpenmrsFetchError } from "@openmrs/esm-api";
import { ComponentContext } from "./ComponentContext.js";
const defaultOpts = {
    strictMode: true,
    throwErrorsToConsole: true,
    disableTranslations: false
};
const swrCache = new Map();
// Read more about the available config options here: https://swr.vercel.app/docs/api#configuration
const defaultSwrConfig = {
    // max number of retries after requests have failed
    errorRetryCount: 3,
    // default fetcher function
    fetcher: openmrsFetch,
    // only revalidate once every 30 minutes
    focusThrottleInterval: 1800000,
    revalidateIfStale: true,
    // disable automatic revalidations by default
    revalidateOnFocus: false,
    revalidateOnReconnect: false,
    refreshInterval: 0,
    shouldRetryOnError: (error)=>{
        if (error instanceof OpenmrsFetchError) {
            const status = error.response.status;
            // retry all server-side errors
            if (status >= 500) {
                return true;
            }
            // retry on authentication failure or rate-limited requests
            if (status === 401 || status === 403 || status === 429) {
                return true;
            }
            // do not retry on other client-side errors
            return false;
        }
        return true;
    },
    provider: ()=>swrCache
};
export function openmrsComponentDecorator(userOpts) {
    if (typeof userOpts !== 'object' || typeof userOpts.featureName !== 'string' || typeof userOpts.moduleName !== 'string') {
        throw new Error('Invalid options');
    }
    const opts = Object.assign({}, defaultOpts, userOpts);
    const swrConfig = {
        ...defaultSwrConfig,
        ...opts.swrConfig
    };
    return function decorateComponent(Comp) {
        var _React_Component, _OpenmrsReactComponent;
        return _OpenmrsReactComponent = class OpenmrsReactComponent extends (_React_Component = React.Component) {
            componentDidCatch(err, info) {
                if (info && info.componentStack) {
                    err.extra = Object.assign(err.extra || {}, {
                        componentStack: info.componentStack
                    });
                }
                if (opts.throwErrorsToConsole) {
                    setTimeout(()=>{
                        throw err;
                    });
                }
                this.setState({
                    caughtError: err,
                    caughtErrorInfo: info
                });
            }
            render() {
                if (this.state.caughtError) {
                    // TO-DO have a UX designed for when a catastrophic error occurs
                    return /*#__PURE__*/ React.createElement("div", null, "An error has occurred. Please try reloading the page.");
                } else {
                    const content = /*#__PURE__*/ React.createElement(Suspense, {
                        fallback: null
                    }, /*#__PURE__*/ React.createElement(SWRConfig, {
                        value: swrConfig
                    }, /*#__PURE__*/ React.createElement(ComponentContext.Provider, {
                        value: this.state.config
                    }, opts.disableTranslations ? /*#__PURE__*/ React.createElement(Comp, this.props) : /*#__PURE__*/ React.createElement(I18nextProvider, {
                        i18n: window.i18next,
                        defaultNS: this.state.config.extension ? `${this.state.config.moduleName}___${this.state.config.extension.extensionSlotName}___${this.state.config.extension.extensionId}` : this.state.config.moduleName
                    }, /*#__PURE__*/ React.createElement(Comp, this.props)))));
                    if (!opts.strictMode || !React.StrictMode) {
                        return content;
                    } else {
                        return /*#__PURE__*/ React.createElement(React.StrictMode, null, content);
                    }
                }
            }
            constructor(props){
                super(props);
                this.state = {
                    caughtError: null,
                    caughtErrorInfo: null,
                    config: {
                        moduleName: opts.moduleName,
                        featureName: opts.featureName,
                        extension: props._extensionContext
                    }
                };
            }
        }, _define_property(_OpenmrsReactComponent, "displayName", `OpenmrsReactComponent(${opts.featureName})`), _OpenmrsReactComponent;
    };
}
