// @openmrs/esm-react-utils/dist/useOpenmrsInfinite.js
/** @module @category UI */ import { openmrsFetch } from "@openmrs/esm-api";
import { useCallback } from "react";
import useSWRInfinite from "swr/infinite";
import { openmrsServerPaginationHandlers } from "./useOpenmrsPagination.js";
/**
 * Most REST endpoints that return a list of objects, such as getAll or search, are server-side paginated.
 * The server limits the max number of results being returned, and multiple requests are needed to get the full data set
 * if its size exceeds this limit.
 * The max number of results per request is configurable server-side
 * with the key "webservices.rest.maxResultsDefault". See: https://openmrs.atlassian.net/wiki/spaces/docs/pages/25469882/REST+Module
 *
 * This hook fetches data from a paginated rest endpoint, initially by fetching the first page of the results.
 * It provides a callback to load data from subsequent pages as needed. This hook is intended to serve UIs that
 * provide infinite loading / scrolling of results. Unlike `useOpenmrsPagination`, this hook does not allow random access
 * (and lazy-loading) of any arbitrary page; rather, it fetches pages sequentially starting form the initial page, and the next page
 * is fetched by calling `loadMore`. See: https://swr.vercel.app/docs/pagination#useswrinfinite
 *
 * @see `useOpenmrsPagination`
 * @see `useOpenmrsFetchAll`
 * @see `useFhirInfinite`
 *
 * @param url The URL of the paginated rest endpoint. Note that the `limit` GET param can be set to specify
 *            the page size; if not set, the page size defaults to the `webservices.rest.maxResultsDefault` value defined
 *            server-side.
 *            Similar to useSWRInfinite, this param can be null to disable fetching.
 * @param options The options object
 * @returns a UseServerInfiniteReturnObject object
 */ export function useOpenmrsInfinite(url, options = {}) {
    return useServerInfinite(url, openmrsServerPaginationHandlers, options);
}
export function useServerInfinite(url, serverPaginationHandlers, options = {}) {
    const { swrInfiniteConfig, immutable } = options;
    const { getNextUrl, getTotalCount, getData } = serverPaginationHandlers;
    const fetcher = options.fetcher ?? openmrsFetch;
    const getKey = useCallback((pageIndex, previousPageData)=>{
        if (pageIndex == 0) {
            return url?.toString() ?? null;
        } else {
            return serverPaginationHandlers.getNextUrl(previousPageData.data);
        }
    }, [
        url
    ]);
    const { data, size, setSize, ...rest } = useSWRInfinite(getKey, fetcher, {
        ...swrInfiniteConfig,
        // `useSWR` has a useSWRImmutable counterpart, but `useSWRInfinite` does not seem to.
        // Setting the revalidate params manually if immutable is true, see: https://swr.vercel.app/docs/revalidation
        ...immutable ? {
            revalidateIfStale: false,
            revalidateOnFocus: false,
            revalidateOnReconnect: false
        } : {}
    });
    const nextUri = getNextUrl(data?.[data.length - 1].data);
    const hasMore = nextUri != null;
    const loadMore = ()=>{
        setSize((data?.length ?? 0) + 1);
    };
    const totalCount = getTotalCount(data?.[0]?.data);
    return {
        data: data?.flatMap((d)=>getData(d.data)),
        totalCount,
        hasMore,
        loadMore,
        nextUri,
        ...rest
    };
}
