// @openmrs/esm-react-utils/dist/useRenderableExtensions.js
/** @module @category Extension */ import React from "react";
import { ComponentContext, Extension, useExtensionSlot } from "./index.js";
/**
 * This is an advanced hook for use-cases where its useful to use the extension system,
 * but not the `ExtensionSlot` component's rendering of extensions. Use of this hook
 * should be avoided if possible.
 *
 * Functionally, this hook is very similar to the `ExtensionSlot` component, but whereas
 * an `ExtensionSlot` renders a DOM tree of extensions bound to the slot, this hook simply
 * returns the extensions as an array of React components that can be wired into a component
 * however makes sense.
 *
 * @param name The name of the extension slot
 *
 * @example
 * ```ts
 * const extensions = useRenderableExtensions('my-extension-slot');
 * return (
 *  <>
 *    {extensions.map((Ext, index) => (
 *      <React.Fragment key={index}>
 *        <Ext state={{key: 'value'}} />
 *      </React.Fragment>
 *    ))}
 *  </>
 * )
 * ```
 */ export function useRenderableExtensions(name) {
    const { extensions, extensionSlotModuleName } = useExtensionSlot(name);
    return name ? extensions.map((extension)=>({ state })=>/*#__PURE__*/ React.createElement(ComponentContext.Provider, {
                value: {
                    moduleName: extensionSlotModuleName,
                    featureName: '',
                    extension: {
                        extensionId: extension.id,
                        extensionSlotName: name,
                        extensionSlotModuleName
                    }
                }
            }, /*#__PURE__*/ React.createElement(Extension, {
                state: state
            }))) : [];
}
