// openmrs/dist/utils/debugger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@rspack/core");
const dev_server_1 = require("@rspack/dev-server");
const node_path_1 = require("node:path");
const webpack_1 = __importDefault(require("webpack"));
const webpack_dev_server_1 = __importDefault(require("webpack-dev-server"));
const logger_1 = require("./logger");
function getWebpackEnv() {
    return {
        ...process.env,
        analyze: process.env.BUNDLE_ANALYZE === 'true',
        production: process.env.NODE_ENV === 'production',
        development: process.env.NODE_ENV === 'development',
    };
}
function loadConfig(configPath) {
    const content = require(configPath);
    if (typeof content === 'function') {
        return content(getWebpackEnv());
    }
    return content;
}
function startDevServer(configPath, port, useRspack = false) {
    const config = loadConfig(configPath);
    const devServerOptions = {
        ...config.devServer,
        port,
        static: (0, node_path_1.dirname)(configPath),
    };
    let server;
    if (!useRspack) {
        const compiler = (0, webpack_1.default)(config);
        server = new webpack_dev_server_1.default(devServerOptions, compiler);
    }
    else {
        const compiler = (0, core_1.rspack)(config);
        server = new dev_server_1.RspackDevServer(devServerOptions, compiler);
    }
    server.startCallback(() => {
        (0, logger_1.logInfo)(`Listening at http://localhost:${port}`);
    });
}
process.on('message', ({ source, port, useRspack }) => startDevServer(source, port, useRspack));
