// @openmrs/esm-emr-api/dist/visit-utils.js
/** @module @category API */ import { openmrsFetch, restBaseUrl } from "@openmrs/esm-api";
import { getGlobalStore } from "@openmrs/esm-state";
import { BehaviorSubject } from "rxjs";
export var VisitMode = /*#__PURE__*/ function(VisitMode) {
    VisitMode["NEWVISIT"] = "startVisit";
    VisitMode["EDITVISIT"] = "editVisit";
    VisitMode["LOADING"] = "loadingVisit";
    return VisitMode;
}({});
export var VisitStatus = /*#__PURE__*/ function(VisitStatus) {
    VisitStatus["NOTSTARTED"] = "notStarted";
    VisitStatus["ONGOING"] = "ongoing";
    return VisitStatus;
}({});
export const defaultVisitCustomRepresentation = 'custom:(uuid,display,voided,indication,startDatetime,stopDatetime,' + 'encounters:(uuid,display,encounterDatetime,' + 'form:(uuid,name),location:ref,' + 'encounterType:ref,' + 'encounterProviders:(uuid,display,' + 'provider:(uuid,display))),' + 'patient:(uuid,display),' + 'visitType:(uuid,name,display),' + 'attributes:(uuid,display,attributeType:(name,datatypeClassname,uuid),value),' + 'location:(uuid,name,display))';
const initialState = getVisitSessionStorage() || {
    patientUuid: null,
    manuallySetVisitUuid: null,
    mutateVisitCallbacks: {}
};
export function getVisitStore() {
    return getGlobalStore('visit', initialState);
}
export function setCurrentVisit(patientUuid, visitUuid) {
    getVisitStore().setState({
        patientUuid,
        manuallySetVisitUuid: visitUuid
    });
}
getVisitStore().subscribe((state)=>{
    setVisitSessionStorage(state);
});
function setVisitSessionStorage(value) {
    sessionStorage.setItem('openmrs:visitStoreState', JSON.stringify(value));
}
function getVisitSessionStorage() {
    try {
        return JSON.parse(sessionStorage.getItem('openmrs:visitStoreState') || 'null');
    } catch (e) {
        return null;
    }
}
export function saveVisit(payload, abortController) {
    return openmrsFetch(`${restBaseUrl}/visit`, {
        signal: abortController.signal,
        method: 'POST',
        headers: {
            'Content-type': 'application/json'
        },
        body: payload
    });
}
export function updateVisit(uuid, payload, abortController) {
    return openmrsFetch(`${restBaseUrl}/visit/${uuid}`, {
        signal: abortController.signal,
        method: 'POST',
        headers: {
            'Content-type': 'application/json'
        },
        body: payload
    });
}
/**
 * @deprecated Use the `useVisit` hook instead.
 */ export function getVisitsForPatient(patientUuid, abortController, v) {
    const custom = v ?? defaultVisitCustomRepresentation;
    return openmrsFetch(`${restBaseUrl}/visit?patient=${patientUuid}&v=${custom}`, {
        signal: abortController.signal,
        method: 'GET',
        headers: {
            'Content-type': 'application/json'
        }
    });
}
/** @deprecated */ export const getStartedVisit = new BehaviorSubject(null);
